### DISCUSSION TITLE
FreeCAD link object

### DISCUSSION THREAD

**Discussion by macdroid53** [nortikin/sverchok#4535] [General]
FreeCAD has a process called Link. It allows solids or whole models to linked into a FreeCAD FCStd file.

Essentially, and reference to external data. These references are apparently not seen by Read FCStd node.

Is it possible to support that via the Read FCStd node?

**Comment by zeffii** [comment]
if it's available from the API, then probably :)

**Comment by macdroid53** [comment]
I'm pretty sure the API does expose it...but the Read FCStd node doesn't see it...

Here WithLink.FCStd has a link to Block.FCStd. But LinkCheck.blend Read FCStd with file path specified to WithLink.FCStd only the body local to WIthLink.FCStd shows up.
[LinkCheck.zip](https://github.com/nortikin/sverchok/files/8921334/LinkCheck.zip)

**Reply by zeffii** [reply]
@macdroid53  well, if the node isn't programmed to look for links (if the API is present) then yeah, they won't show up :)

cool, thanks for the linkcheck zip :) i'll poke around

**Comment by zeffii** [comment]
ok good!
```python
>>> import FreeCAD as F
>>> fn = r"D:\DOWNLOADS\2022\fcstd_files\LinkCheck\WithLink.FCStd"
>>> doc = F.open(fn)

>>> doc.OutList
[<Document object at 000001C3DD6E4D90>]
>>> [i.Label for i in doc.OutList]
['Block']

>>> block = doc.OutList[0]
>>> block.FileName
'D:/DOWNLOADS/2022/fcstd_files/LinkCheck/Block.FCStd'

>>> for obj in doc.Objects:
...     if obj.TypeId == "App::Link":
...         shape = obj.LinkedObject.Shape
...         print(shape.tessellate(1.0))
... 
([Vector (26.672863, 20.07435, 10.0), Vector (-23.141266, 20.074350000000003, 0.0), Vector (-23.141266, 20.074350000000003, 10.0), Vector (26.672863, 20.07435, 0.0), Vector (26.672863, -21.747211, 0.0), Vector (26.672863, -21.747211, 10.0), Vector (-23.141266, -21.747211, 10.0), Vector (-23.141266, -21.747211, 0.0)], [(0, 1, 2), (3, 1, 0), (4, 3, 0), (4, 0, 5), (6, 4, 5), (7, 4, 6), (1, 7, 6), (1, 6, 2), (1, 3, 7), (7, 3, 4), (0, 2, 6), (0, 6, 5)])
```

now some of the commented out code in the fcstd_read_mod.py code makes sense.

https://github.com/nortikin/sverchok/pull/4534/files#diff-5cb4f5bb15ac364393b081451e91ec70b617b0b4732540d1694595e2e931d144R240-R254

**Reply by zeffii** [reply]
the `MOD` version of the node does already import links.. not sure how deeply nested that would work..
![image](https://user-images.githubusercontent.com/619340/174262406-a036300b-a11e-4193-aed0-2c61e79a6609.png)

**Reply by macdroid53** [reply]
Is the MOD version in a branch? Or does "check for updates" get that?

**Reply by zeffii** [reply]
i committed a few hours ago :) you'll see it.
